src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,73 @@
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryDiscardDesignation
+{
+    /// <summary>
+    /// Supports code like <c>o switch { int _ => ... }</c> to just <c>o switch { int => ... }</c> in C# 9 and above.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.RemoveUnnecessaryDiscardDesignationDiagnosticId,
+                   option: null,
+                   new LocalizableResourceString(nameof(CSharpAnalyzersResources.Remove_unnessary_discard), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+                   new LocalizableResourceString(nameof(CSharpAnalyzersResources.Discard_can_be_removed), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+                   isUnnecessary: true)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeDiscardDesignation, SyntaxKind.DiscardDesignation);
+        }
+
+        private void AnalyzeDiscardDesignation(SyntaxNodeAnalysisContext context)
+        {
+            var syntaxTree = context.SemanticModel.SyntaxTree;
+            if (((CSharpParseOptions)syntaxTree.Options).LanguageVersion < LanguageVersion.CSharp9)
+                return;
+
+            var discard = (DiscardDesignationSyntax)context.Node;
+            if (discard.Parent is DeclarationPatternSyntax)
+            {
+                Report(discard);
+            }
+            else if (discard.Parent is RecursivePatternSyntax recursivePattern)
+            {
+                // can't remove from `(int i) _` as `(int i)` is not a legal pattern itself.
+                if (recursivePattern.PositionalPatternClause != null &&
+                    recursivePattern.PositionalPatternClause.Subpatterns.Count == 1)
+                {
+                    return;
+                }
+
+                Report(discard);
+            }
+
+            return;
+
+            void Report(DiscardDesignationSyntax discard)
+            {
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    this.Descriptor,
+                    discard.GetLocation(),
+                    ReportDiagnostic.Default,
+                    SpecializedCollections.EmptyEnumerable<Location>(),
+                    ImmutableDictionary<string, string>.Empty));
+            }
```

Comment:
```suggestion
            void Report(DiscardDesignationSyntax discard)
                => context.ReportDiagnostic(Diagnostic.Create(Descriptor, discard.GetLocation());
```
```

Reply:
What is DiagnosticHelper for?

Reply:
> What is DiagnosticHelper for?

For analyzers with code style options, as option itself has a notification/severity that needs to be considered for the effective severity of reported diagnostic. For analyzers without code style option, it should just use the simpler public API to create diagnostics.

Reply:
done, thanks.